sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+  az_json_builder json_builder;
+  AZ_RETURN_IF_FAILED(az_json_builder_init(&json_builder, response, NULL));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_begin_object(&json_builder));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_property_name(&json_builder, report_max_temp_name_span));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_int32_number(&json_builder, (int32_t)device_max_temp));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_property_name(&json_builder, report_min_temp_name_span));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_int32_number(&json_builder, (int32_t)device_min_temp));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_property_name(&json_builder, report_avg_temp_name_span));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_int32_number(&json_builder, (int32_t)device_avg_temp));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_property_name(&json_builder, report_start_time_name_span));
+
+  // If the user specified a time, use that as the start | Otherwise use boot time
+  if (az_span_size(start_time_span) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_json_builder_append_string(&json_builder, start_time_span));
+  }
+  else
+  {
+    AZ_RETURN_IF_FAILED(az_json_builder_append_string(&json_builder, boot_time_span));
+  }
+
+  AZ_RETURN_IF_FAILED(az_json_builder_append_property_name(&json_builder, report_end_time_name_span));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_string(&json_builder, time_span));
+  AZ_RETURN_IF_FAILED(az_json_builder_append_end_object(&json_builder));
+
+  *out_response = az_json_builder_get_json(&json_builder);
+
+  return AZ_OK;
+}
+
+// Send the response of the method invocation
+static int send_method_response(
+    az_iot_hub_client_method_request* request,
+    uint16_t status,
+    az_span response)
+{
+  // Get the response topic to publish the method response
+  if (az_failed(az_iot_hub_client_methods_response_get_publish_topic(
+          &client,
+          request->request_id,
+          status,
+          methods_response_topic,
+          sizeof(methods_response_topic),
+          NULL)))
+  {
+    printf("Unable to get twin document publish topic");
+    return -1;
+  }
+
+  printf("Status: %u\tPayload:", status);
+  char* payload_char = (char*)az_span_ptr(response);
+  if (payload_char != NULL)
+  {
+    for (int32_t i = 0; i < az_span_size(response); i++)
+    {
+      putchar(*(payload_char + i));
+    }
+    putchar('\n');
+  }
+
+  // Send the methods response
+  int rc;
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           methods_response_topic,
+           az_span_size(response),
+           az_span_ptr(response),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to send method response, return code %d\n", rc);
+    return rc;
+  }
+
+  printf("Sent response\n");
+
+  return 0;
+}
+
+// Build the JSON payload for the reported property
+static az_result build_reported_property(az_json_builder* json_builder, double property_val)
```

Comment:
In `send_reported_temperature_property` you're invoking this function with `desired_temp` but it looks like we're setting the property name to be `max_temp` here always?

Also, when responding to a property request from server its moer complicated in 1st place.  We need the value we're using to be wrapped in a `value` envelope and the version (you need to parse $version from desired to get this) so that server, which is distributed sysetm, can know which $version of its send the device last processed.  And also an hTTP status code (e.g. if desiredTemp=400, then a thermostat would respond with its own status400:)) and optionally description.

We can talk more but from spec the canonical example of a response property:
```json
      "goose": {
        "value": "canadian", // Valid P&P property value (embedded value (string))
        "ac": 200, // ackCode for “goose” property 
        "av": 3, // ackVersion for “goose” property
        "ad": "OK" // ackDescription for “goose” property
      }

```

Reply:
My understanding from this section of the DTMI:
https://github.com/Azure/opendigitaltwins-dtdl/blob/531767f1166f2a32429f676e7743f6b418376f2b/DTDL/v2/samples/Thermostat.json#L20-L40
 Is that the incoming, desired property is `targetTemperature` which is parsed and set to the device's current temp. The reported property is `maxTempSinceLastReboot` which doesn't necessarily have anything to do with a target temperature (meaning relative to a response). To make the cadence of sending a reported property consistent, I send it whenever a desired property comes in.
We in theory (I believe) could send this reported property at an interval kind of like telemetry.

Am I misunderstanding twin here? Is there a response other than the MQTT ack that it received the desired prop?

Reply:
Ultimately the "ac"/"av" stuff isn't about the DTDL which is a contrcat at data layer.  This comes from the IoTHub binding, as in what does it take to make a device being PNP to a Hub with deviceClient?  For this the metadata is needed for this.

For the `maxTempSinceLastReboot` I'm fine if this is sent on everytime we respond to the temp, just for a simpler sample.  And this does NOT need metadata since it's "reoadonly" from server side.  But for responding to the `targetTempetarue` does need metadata.

Reply:
I'll make a note to ask you about this but this part
> But for responding to the `targetTemperature` does need metadata
As I currently understand there is no "respond" from this. Is it part of the spec that any desired property will have an according reported property? That seems like a user decision? Especially since in this case the telemetry is essentially doing that job.

Reply:
done